itest/lndsigner_test.go

```diff
@@ -0,0 +1,840 @@
+		synced = 0
+
+		for synced != blocks {
+			time.Sleep(100 * time.Millisecond)
+
+			resp = lnd.Lncli("getinfo")
+			require.Equal(tctx.t, resp["identity_pubkey"].(string),
+				lnd.idPubKey)
+
+			synced = int(resp["block_height"].(float64))
+		}
+	}
+}
+
+// waitForGraphSync ensures that each LND has a synchronized graph.
+func (tctx *testContext) waitForGraphSync() {
+	tctx.t.Helper()
+
+	var (
+		nodes, chans int
+		synced       bool
+		resp         map[string]interface{}
+	)
+
+	tctx.lndMtx.RLock()
+	defer tctx.lndMtx.RUnlock()
+
+	for !synced {
+		synced = true
+
+		for _, lnd := range tctx.lnds {
+
+			resp = lnd.Lncli("getnetworkinfo")
+
+			gotNodes := int(resp["num_nodes"].(float64))
+			if gotNodes != nodes {
+				synced = false
+
+				if gotNodes > nodes {
+					nodes = gotNodes
+				}
+			}
+
+			gotChans := int(resp["num_channels"].(float64))
+			if gotChans != chans {
+				synced = false
+
+				if gotChans > chans {
+					chans = gotChans
+				}
+			}
+		}
+
+		time.Sleep(500 * time.Millisecond)
+	}
+}
+
+// testEach runs a function for each LND instance, in parallel.
+func (tctx *testContext) testEach(test func(lnd *lndHarness)) {
+	tctx.t.Helper()
+
+	tctx.lndMtx.RLock()
+	defer tctx.lndMtx.RUnlock()
+
+	var wg sync.WaitGroup
+	wg.Add(len(tctx.lnds))
+
+	for _, lnd := range tctx.lnds {
+		innerLnd := lnd
+
+		go func() {
+			defer wg.Done()
+
+			test(innerLnd)
+		}()
+	}
+
+	wg.Wait()
+}
+
+// testEachPair runs a function for each pair of LNDs in parallel, avoiding
+// testing an LND instance with itself.
+func (tctx *testContext) testEachPair(test func(lnd1, lnd2 *lndHarness)) {
+	tctx.t.Helper()
+
+	tctx.lndMtx.RLock()
+	defer tctx.lndMtx.RUnlock()
+
+	var wg sync.WaitGroup
+	wg.Add(len(tctx.lnds)*len(tctx.lnds) - len(tctx.lnds))
+
+	for i, lnd1 := range tctx.lnds {
+		for j, lnd2 := range tctx.lnds {
+			innerLnd1 := lnd1
+			innerLnd2 := lnd2
+
+			if i == j {
+				continue
+			}
+
+			go func() {
+				defer wg.Done()
+
+				test(innerLnd1, innerLnd2)
+			}()
+		}
+	}
+
+	wg.Wait()
+}
+
+// lndHarness manages a single lndsignerd-backed instance of LND.
+type lndHarness struct {
+	tctx     *testContext
+	idPubKey string
+
+	unixSocket bool
+
+	lndSignerCmd *exec.Cmd
+	lndSignerMtx sync.RWMutex
+
+	lndDir    string
+	lncliPath string
+	rpc       string
+	p2p       string
+	lndCmd    *exec.Cmd
+	lndMtx    sync.RWMutex
+}
+
+// Start takes the initial configuration (tctx, idPubKey, and unixSocket) and
+// starts lndsignerd and LND.
+func (l *lndHarness) Start() {
+	l.tctx.t.Helper()
+
+	// Start lndsignerd.
+	l.lndDir = path.Join(l.tctx.tmpRoot, fmt.Sprintf("lnd%s", l.idPubKey))
+	err := os.Mkdir(l.lndDir, fs.ModeDir|0700)
+	require.NoError(l.tctx.t, err)
+
+	signerAddr := "127.0.0.1:" + newPortString()
+	fullSignerAddr := "tcp://" + signerAddr
+
+	if l.unixSocket {
+		signerAddr := path.Join(l.lndDir, "signer.socket")
+		fullSignerAddr = "unix://" + signerAddr
+	}
+
+	l.lndSignerMtx.Lock()
+	defer l.lndSignerMtx.Unlock()
```

Comment:
The instance is created in `addNode()` and `Start()` is called immediately after, so it shouldn't be in use. The mutex is, again to avoid race detection complaining despite the fact that a true race won't happen here.

Reply:
Maybe you can let the goroutine push the instance variable to a channel, and store them without a lock in the main thread? Those mutexes do raise questions when reviewing the code.

Reply:
I have simplified this greatly in https://github.com/bottlepay/lndsigner/compare/66c3baaad33488e5b25d2e8af0935f3adbec2880..d7ba0804a5348e157f205542fa403bf371d91d80 and eliminated the need for mutexes.